File: clients/windows/src/ClipShare.Windows.Vault/VaultEvents.cs
namespace ClipShare.Windows.Vault;

using System.Security.Cryptography;
using System.Text;

public enum EntityKind
{
    Folder,
    Item,
}

public enum VaultOperation
{
    Create,
    Update,
    Move,
    Trash,
    Restore,
    Purge,
}

public sealed record VaultEvent(
    EventId EventId,
    DeviceId SourceDeviceId,
    long SourceSequence,
    long LogicalTime,
    EntityKind EntityKind,
    string EntityId,
    VaultOperation Operation,
    VersionId? BaseVersionId,
    VersionId NewVersionId,
    long KeyEpoch,
    string EncryptedBodyDigest)
{
    public VaultEvent Validate()
    {
        if (SourceSequence <= 0 || LogicalTime <= 0 || KeyEpoch <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(SourceSequence), "Event counters and epoch are positive.");
        }

        if (string.IsNullOrEmpty(EncryptedBodyDigest) || EncryptedBodyDigest.Length != 64 || EncryptedBodyDigest.Any(character =>
            !Uri.IsHexDigit(character) || character is >= 'A' and <= 'F'))
        {
            throw new ArgumentException("Event body digest must be lowercase SHA-256.", nameof(EncryptedBodyDigest));
        }

        if (!Enum.IsDefined(EntityKind) || !Enum.IsDefined(Operation))
        {
            throw new ArgumentOutOfRangeException(nameof(EntityKind), "Unknown Vault event kind or operation.");
        }

        _ = VaultUuid.Validate(EventId.Value);
        _ = VaultUuid.Validate(SourceDeviceId.Value);
        _ = VaultUuid.Validate(EntityId);
        _ = VaultUuid.Validate(NewVersionId.Value);
        if (BaseVersionId is { } baseVersionId)
        {
            _ = VaultUuid.Validate(baseVersionId.Value);
        }
        return this;
    }
}

public sealed record EntitySnapshot(VersionId? CurrentVersionId, bool Tombstone, bool MoveWouldCreateCycle = false);

public abstract record EventDecision
{
    private EventDecision()
    {
    }

    public sealed record ApplyLinear : EventDecision;

    public sealed record Duplicate : EventDecision;

    public sealed record CreateConflictCopy(string ConflictEntityId) : EventDecision;

    public sealed record PreserveTombstoneAndConflict(string ConflictEntityId) : EventDecision;

    public sealed record QuarantineCycle : EventDecision;
}

public sealed class EventRejectedException(string message) : InvalidOperationException(message);

public sealed class VaultEventLedger
{
    private readonly IReadOnlySet<long> acceptedEpochs;
    private readonly IReadOnlySet<DeviceId> revokedDevices;
    private readonly Dictionary<EventId, VaultEvent> byEventId = [];
    private readonly Dictionary<(DeviceId Device, long Sequence), EventId> bySourceSequence = [];
    private readonly Dictionary<DeviceId, long> highestSourceSequence = [];

    public VaultEventLedger(IReadOnlySet<long> acceptedEpochs, IReadOnlySet<DeviceId>? revokedDevices = null)
    {
        ArgumentNullException.ThrowIfNull(acceptedEpochs);
        this.acceptedEpochs = acceptedEpochs;
        this.revokedDevices = revokedDevices ?? new HashSet<DeviceId>();
    }

    public EventDecision Accept(VaultEvent vaultEvent, EntitySnapshot snapshot)
    {
        ArgumentNullException.ThrowIfNull(vaultEvent);
        ArgumentNullException.ThrowIfNull(snapshot);
        vaultEvent.Validate();
        if (byEventId.TryGetValue(vaultEvent.EventId, out var existing))
        {
            return existing == vaultEvent
                ? new EventDecision.Duplicate()
                : throw new EventRejectedException("An event ID was reused with different content.");
        }

        if (revokedDevices.Contains(vaultEvent.SourceDeviceId))
        {
            throw new EventRejectedException("Events from revoked devices are rejected.");
        }

        if (!acceptedEpochs.Contains(vaultEvent.KeyEpoch))
        {
            throw new EventRejectedException("Events for unavailable epochs are rejected.");
        }

        var sourceKey = (vaultEvent.SourceDeviceId, vaultEvent.SourceSequence);
        if (bySourceSequence.ContainsKey(sourceKey))
        {
            throw new EventRejectedException("A source sequence was reused by a different event.");
        }

        if (highestSourceSequence.TryGetValue(vaultEvent.SourceDeviceId, out long highest) &&
            vaultEvent.SourceSequence <= highest)
        {
            throw new EventRejectedException("An older source sequence cannot be accepted after a newer event.");
        }

        var decision = Decide(vaultEvent, snapshot);
        byEventId[vaultEvent.EventId] = vaultEvent;
        bySourceSequence[sourceKey] = vaultEvent.EventId;
        highestSourceSequence[vaultEvent.SourceDeviceId] = vaultEvent.SourceSequence;
        return decision;
    }

    public static string ConflictEntityId(string entityId, EventId eventId)
    {
        var prefix = Encoding.ASCII.GetBytes("clipshare:vault:conflict:v1\0");
        var entity = Convert.FromHexString(VaultUuid.Validate(entityId).Replace("-", string.Empty, StringComparison.Ordinal));
        var eventBytes = Convert.FromHexString(eventId.Value.Replace("-", string.Empty, StringComparison.Ordinal));
        var digest = SHA256.HashData(prefix.Concat(entity).Concat(eventBytes).ToArray());
        var bytes = digest.AsSpan(0, 16).ToArray();
        bytes[6] = (byte)((bytes[6] & 0x0f) | 0x80);
        bytes[8] = (byte)((bytes[8] & 0x3f) | 0x80);
        var hex = Convert.ToHexString(bytes).ToLowerInvariant();
        return $"{hex[..8]}-{hex[8..12]}-{hex[12..16]}-{hex[16..20]}-{hex[20..]}";
    }

    private static EventDecision Decide(VaultEvent vaultEvent, EntitySnapshot snapshot)
    {
        if (vaultEvent.BaseVersionId == snapshot.CurrentVersionId)
        {
            return new EventDecision.ApplyLinear();
        }

        if (vaultEvent.EntityKind == EntityKind.Folder &&
            vaultEvent.Operation == VaultOperation.Move &&
            snapshot.MoveWouldCreateCycle)
        {
            return new EventDecision.QuarantineCycle();
        }

        var conflictId = ConflictEntityId(vaultEvent.EntityId, vaultEvent.EventId);
        return vaultEvent.Operation == VaultOperation.Trash || snapshot.Tombstone
            ? new EventDecision.PreserveTombstoneAndConflict(conflictId)
            : new EventDecision.CreateConflictCopy(conflictId);
    }
}

public static class TrashRetention
{
    public static TimeSpan Duration { get; } = TimeSpan.FromDays(30);

    public static bool IsExpired(DateTimeOffset deletedAt, DateTimeOffset now) => now >= deletedAt + Duration;
}

public enum ReencryptionState
{
    Plan,
    Copying,
    Paused,
    Verifying,
    Committing,
    Complete,
    FailedRollback,
}

public sealed record HistoricalReencryptionJob
{
    public HistoricalReencryptionJob(
        long oldEpoch,
        long newEpoch,
        long totalRecords,
        long processedRecords = 0,
        ReencryptionState state = ReencryptionState.Plan)
    {
        if (oldEpoch <= 0 || newEpoch <= oldEpoch || totalRecords < 0 || processedRecords < 0 || processedRecords > totalRecords)
        {
            throw new ArgumentOutOfRangeException(nameof(oldEpoch), "Invalid historical reencryption plan.");
        }

        OldEpoch = oldEpoch;
        NewEpoch = newEpoch;
        TotalRecords = totalRecords;
        ProcessedRecords = processedRecords;
        State = state;
    }

    public long OldEpoch { get; init; }

    public long NewEpoch { get; init; }

    public long TotalRecords { get; init; }

    public long ProcessedRecords { get; init; }

    public ReencryptionState State { get; init; }

    public HistoricalReencryptionJob Start() => Transition(ReencryptionState.Plan, ReencryptionState.Copying);

    public HistoricalReencryptionJob Copied(long records)
    {
        if (State != ReencryptionState.Copying || records <= 0 || ProcessedRecords > TotalRecords - records)
        {
            throw new InvalidOperationException("Invalid reencryption copy progress.");
        }

        return this with { ProcessedRecords = ProcessedRecords + records };
    }

    public HistoricalReencryptionJob Pause() => Transition(ReencryptionState.Copying, ReencryptionState.Paused);

    public HistoricalReencryptionJob Resume() => Transition(ReencryptionState.Paused, ReencryptionState.Copying);

    public HistoricalReencryptionJob Verify()
    {
        if (State != ReencryptionState.Copying || ProcessedRecords != TotalRecords)
        {
            throw new InvalidOperationException("All records must be copied before verification.");
        }

        return this with { State = ReencryptionState.Verifying };
    }

    public HistoricalReencryptionJob Commit() => Transition(ReencryptionState.Verifying, ReencryptionState.Committing);

    public HistoricalReencryptionJob Complete(long oldEpochReferenceCount)
    {
        if (State != ReencryptionState.Committing || oldEpochReferenceCount != 0)
        {
            throw new InvalidOperationException("Old epoch references must be zero before completion.");
        }

        return this with { State = ReencryptionState.Complete };
    }

    public HistoricalReencryptionJob FailRollback() =>
        State is not (ReencryptionState.Complete or ReencryptionState.FailedRollback)
            ? this with { State = ReencryptionState.FailedRollback }
            : throw new InvalidOperationException("Completed or already failed reencryption jobs are terminal.");

    private HistoricalReencryptionJob Transition(ReencryptionState from, ReencryptionState to) =>
        State == from
            ? this with { State = to }
            : throw new InvalidOperationException("Invalid reencryption state transition.");
}

File: clients/android/core/sync/src/main/kotlin/com/clipshare/core/sync/VaultEvents.kt
@file:Suppress("MagicNumber", "ReturnCount", "ThrowsCount")

package com.clipshare.core.sync

import com.clipshare.core.vault.DeviceId
import com.clipshare.core.vault.EventId
import com.clipshare.core.vault.VersionId
import java.security.MessageDigest
import java.time.Duration
import java.time.Instant

enum class EntityKind {
    FOLDER,
    ITEM,
}

enum class VaultOperation {
    CREATE,
    UPDATE,
    MOVE,
    TRASH,
    RESTORE,
    PURGE,
}

data class VaultEvent(
    val eventId: EventId,
    val sourceDeviceId: DeviceId,
    val sourceSequence: Long,
    val logicalTime: Long,
    val entityKind: EntityKind,
    val entityId: String,
    val operation: VaultOperation,
    val baseVersionId: VersionId?,
    val newVersionId: VersionId,
    val keyEpoch: Long,
    val encryptedBodyDigest: String,
) {
    init {
        require(sourceSequence > 0 && logicalTime > 0 && keyEpoch > 0) { "Event counters and epoch are positive." }
        require(Regex("^[0-9a-f]{64}$").matches(encryptedBodyDigest)) { "Event body digest must be SHA-256." }
        require(LOWERCASE_UUID.matches(entityId)) { "Event entity IDs must be lowercase UUIDs." }
    }
}

data class EntitySnapshot(
    val currentVersionId: VersionId?,
    val tombstone: Boolean,
    val moveWouldCreateCycle: Boolean = false,
)

sealed interface EventDecision {
    data object ApplyLinear : EventDecision

    data object Duplicate : EventDecision

    data class CreateConflictCopy(val conflictEntityId: String) : EventDecision

    data class PreserveTombstoneAndConflict(val conflictEntityId: String) : EventDecision

    data object QuarantineCycle : EventDecision
}

class EventRejectedException(message: String) : IllegalStateException(message)

class VaultEventLedger(
    private val acceptedEpochs: Set<Long>,
    private val revokedDevices: Set<DeviceId> = emptySet(),
) {
    private val byEventId = mutableMapOf<EventId, VaultEvent>()
    private val bySourceSequence = mutableMapOf<Pair<DeviceId, Long>, EventId>()
    private val highestSourceSequence = mutableMapOf<DeviceId, Long>()

    fun accept(event: VaultEvent, snapshot: EntitySnapshot): EventDecision {
        val existing = byEventId[event.eventId]
        if (existing != null) {
            if (existing == event) return EventDecision.Duplicate
            throw EventRejectedException("An event ID was reused with different content.")
        }
        if (event.sourceDeviceId in revokedDevices) {
            throw EventRejectedException("Events from revoked devices are rejected.")
        }
        if (event.keyEpoch !in acceptedEpochs) {
            throw EventRejectedException("Events for unavailable epochs are rejected.")
        }

        val sourceKey = event.sourceDeviceId to event.sourceSequence
        if (bySourceSequence.containsKey(sourceKey)) {
            throw EventRejectedException("A source sequence was reused by a different event.")
        }
        val highest = highestSourceSequence[event.sourceDeviceId]
        if (highest != null && event.sourceSequence <= highest) {
            throw EventRejectedException("An older source sequence cannot be accepted after a newer event.")
        }

        val decision = decide(event, snapshot)
        byEventId[event.eventId] = event
        bySourceSequence[sourceKey] = event.eventId
        highestSourceSequence[event.sourceDeviceId] = event.sourceSequence
        return decision
    }

    private fun decide(event: VaultEvent, snapshot: EntitySnapshot): EventDecision {
        if (event.baseVersionId == snapshot.currentVersionId) return EventDecision.ApplyLinear
        if (event.entityKind == EntityKind.FOLDER &&
            event.operation == VaultOperation.MOVE &&
            snapshot.moveWouldCreateCycle
        ) {
            return EventDecision.QuarantineCycle
        }
        val conflictId = conflictEntityId(event.entityId, event.eventId)
        return if (event.operation == VaultOperation.TRASH || snapshot.tombstone) {
            EventDecision.PreserveTombstoneAndConflict(conflictId)
        } else {
            EventDecision.CreateConflictCopy(conflictId)
        }
    }

    companion object {
        fun conflictEntityId(entityId: String, eventId: EventId): String {
            val digest = MessageDigest.getInstance("SHA-256")
            digest.update("clipshare:vault:conflict:v1\u0000".toByteArray(Charsets.US_ASCII))
            digest.update(uuidBytes(entityId))
            digest.update(uuidBytes(eventId.value))
            val bytes = digest.digest().copyOf(16)
            bytes[6] = ((bytes[6].toInt() and 0x0f) or 0x80).toByte()
            bytes[8] = ((bytes[8].toInt() and 0x3f) or 0x80).toByte()
            val hex = bytes.toLowerHex()
            return buildString {
                append(hex.substring(0, 8), '-')
                append(hex.substring(8, 12), '-')
                append(hex.substring(12, 16), '-')
                append(hex.substring(16, 20), '-')
                append(hex.substring(20))
            }
        }

        private fun uuidBytes(value: String): ByteArray {
            require(Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$").matches(value)) {
                "Invalid lowercase UUID."
            }
            return value.replace("-", "").chunked(2).map { it.toInt(16).toByte() }.toByteArray()
        }

        private fun ByteArray.toLowerHex(): String = joinToString("") { byte ->
            val value = byte.toInt() and 0xff
            "${HEX[value ushr 4]}${HEX[value and 0x0f]}"
        }

        private const val HEX = "0123456789abcdef"
    }
}

object TrashRetention {
    val duration: Duration = Duration.ofDays(30)

    fun isExpired(deletedAt: Instant, now: Instant): Boolean =
        !now.isBefore(deletedAt.plus(duration))
}

enum class ReencryptionState {
    PLAN,
    COPYING,
    PAUSED,
    VERIFYING,
    COMMITTING,
    COMPLETE,
    FAILED_ROLLBACK,
}

data class HistoricalReencryptionJob(
    val oldEpoch: Long,
    val newEpoch: Long,
    val totalRecords: Long,
    val processedRecords: Long = 0,
    val state: ReencryptionState = ReencryptionState.PLAN,
) {
    init {
        require(oldEpoch > 0 && newEpoch > oldEpoch) { "Reencryption must move to a newer independent epoch." }
        require(totalRecords >= 0 && processedRecords in 0..totalRecords) { "Invalid reencryption progress." }
    }

    fun start() = transition(ReencryptionState.PLAN, ReencryptionState.COPYING)

    fun copied(records: Long): HistoricalReencryptionJob {
        require(state == ReencryptionState.COPYING && records > 0) { "Copy progress is only valid while copying." }
        return copy(processedRecords = Math.addExact(processedRecords, records)).also {
            require(it.processedRecords <= totalRecords) { "Copy progress exceeds the plan." }
        }
    }

    fun pause() = transition(ReencryptionState.COPYING, ReencryptionState.PAUSED)

    fun resume() = transition(ReencryptionState.PAUSED, ReencryptionState.COPYING)

    fun verify(): HistoricalReencryptionJob {
        require(state == ReencryptionState.COPYING && processedRecords == totalRecords) {
            "All records must be copied first."
        }
        return copy(state = ReencryptionState.VERIFYING)
    }

    fun commit() = transition(ReencryptionState.VERIFYING, ReencryptionState.COMMITTING)

    fun complete(oldEpochReferenceCount: Long): HistoricalReencryptionJob {
        require(state == ReencryptionState.COMMITTING && oldEpochReferenceCount == 0L) {
            "Old epoch references must be zero before completion."
        }
        return copy(state = ReencryptionState.COMPLETE)
    }

    fun failRollback(): HistoricalReencryptionJob {
        require(state != ReencryptionState.COMPLETE && state != ReencryptionState.FAILED_ROLLBACK) {
            "Completed or already failed reencryption jobs are terminal."
        }
        return copy(state = ReencryptionState.FAILED_ROLLBACK)
    }

    private fun transition(from: ReencryptionState, to: ReencryptionState): HistoricalReencryptionJob {
        require(state == from) { "Invalid reencryption state transition." }
        return copy(state = to)
    }
}

private val LOWERCASE_UUID = Regex(
    "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$",
)
